File: Python/algospot/FENCE.py
import os, sys

C = int(input()) # 테스트 케이스

sys.setrecursionlimit(100000)

results = []

for _ in range(C):
    N = int(sys.stdin.readline()) # 판자 수
    planks = []
    for numStr in sys.stdin.readline().split():
        planks.append(int(numStr))

    def solve(left, right):
        global planks
        if left == right: return planks[left]

        mid = int((left + right) / 2)

        ret = max(solve(left, mid), solve(mid + 1, right))

        low = mid; high = mid + 1
        height = min(planks[low], planks[high])

        ret = max(ret, height * 2)
        
        while left < low or high < right:
            if high < right and (low == left or planks[low - 1] < planks[high + 1]):
                high += 1
                height = min(height, planks[high])
            else:
                low -= 1
                height = min(height, planks[low])
        
            ret = max(ret, height * (high - low + 1))

        return ret

    max_size = solve(0, N - 1)

    results.append(str(max_size))

print('\n'.join(results))

File: CPP/Algospot/FENCE.cpp
#include <iostream>
#include <vector>
#include <functional>

int main() {
    int N;
    std::cin >> N;
    
    for (int _ = 0; _ < N; _++) {
        int numberCount;
        std::cin >> numberCount;
        std::vector<int> planks;
        for (int i = 0; i < numberCount; i++) {
            int num;
            std::cin >> num;
            planks.push_back(num);
        }

        std::function<int(int, int)> solve = [&](int left, int right) -> int {
            if (left == right) return planks.at(left);

            int mid = (left + right) / 2;
            int max_size = std::max(solve(left, mid), solve(mid + 1, right));

            int low = mid, high = mid + 1;
            int height = std::min(planks[low], planks[high]);
            max_size = std::max(max_size, height * 2);

            while (left < low || high < right) {
                if (high < right && (low == left || planks[low - 1] < planks[high + 1])) {
                    high += 1;
                    height = std::min(height, planks[high]);
                } else {
                    low -= 1;
                    height = std::min(height, planks[low]);
                }
            }

            max_size = std::max(max_size, height * (high - low + 1));

            return max_size;
        };

        int max_size = solve(0, numberCount - 1);
        std::cout << max_size << std::endl;
    }

    return 0;
}